### ISSUE TITLE
Make Impl methods `async`?

### ISSUE THREAD

**Issue by kasbah** [hpx7/rtag#190]
Would be nice if all the methods were declared as `async` / returning `Promise<>` so I can `await` in them if needed. As far as I can see they are currently called in async functions in `wrapper.ts.hbs`

What do you think? It will be a breaking change but fairly straight-forward to do.

**Comment by hpx7** [comment]
What use case do you have for awaiting inside methods? A common one I've heard is making http requests inside methods, but in order to do that I plan on adding support for a `ctx.http()` method (since we need to be able to replay the http requests).

The challenge with async methods is that Hathora would need to await them by default to ensure that transactions don't interleave. If your transactions (methods) take time to resolve, this could reduce your app's throughput overall.

**Comment by kasbah** [comment]
My use case is that I need to [call an async method from an external library](https://github.com/kasbah/calm-go/blob/f6aea69c9a5ae579430872ac6a8ec8d838b3d3d4/server/impl.ts#L324). In this case I'd like to do it in `getUserState` but I can easily imagine wanting to do something like that in any of the other methods.

I did realize I'm probably not considering the full issue when I just tacked async on. I do have some questions though:

- Why do you need to be able to replay http methods? 
- What's the exact issue with interleaving transactions, is it that they have shared mutable state?

**Comment by kasbah** [comment]
I was thinking a bit about this issue and if using immutable data for state might solve it. I think the problem would still exist: you are essentially creating two states if you have two async methods running at the same time. You could try and merge them at the end but they may overwrite each other. It's a little less confusing to me when thinking about it with immutable state data but the problem remains. (This short egghead video [on the immer docs](https://immerjs.github.io/immer/async/) actually helped me think this through.)

Thinking about my specific use case. I actually only need async in `getUserState`. Since the state isn't (and shouldn't be) modified in `getUserState` I think it would actually be fine for that to be async. Do you agree? 

I also looked at the `loadState` implementation and realised why you want to be able to replay http methods. That does leave the question of how to handle async effects more generally. Making a special case for http  seems a bit short-sighted. Maybe a more general function to schedule (and later replay) async tasks could be implemented, something that could be added to the `Methods` class as `this.schedule` or just `hathora.schedule`? E.g:

```ts
export class Impl implements Methods<InternalState> {
  // ... 
  joinGame(
    state: InternalState,
    userId: UserId,
    ctx: Context,
    request: IJoinGameRequest
  ): Response {
    state.exampleUrl = `https://example.com/${request.exampleName}`;
    this.schedule(async (state: InternalState) => {
      const data = await fetch(state.exampleUrl).then((r) => r.json());
      state.data = data;
    });
    return Response.ok();
  }
  // ... 
}
```

I acknowledge there is a bit of a problem there in that it's very tempting to use the `state` and the other `joinGame` arguments inside the async callback though.

**Comment by hpx7** [comment]
Yes you are correct in your assessment on what makes async tricky -- the shared mutable state in methods is the `InternalState` (i.e. `state`). And yes even if `state` were immutable you still wouldn't want to interleave transactions.

As a side note -- I have thought about making `state` immutable but I don't think it's possible with the current model. The issue is that `InternalState` isn't necessarily cloneable. For example your internal state includes `board: Board;`, which Hathora/Immerjs wouldn't know how to clone (immerjs only works on js primitives and built in data structures, for custom classes it's a bit dodgy).

I think it's fine to mark all the methods as `async` as you started in your PR, we just have to make sure to await them inside `wrapper.ts`. As long as we await the methods it will guarantee serial execution of the methods. This essentially makes all async calls "blocking" which will reduce throughput if the async calls take some time, but I think that's acceptable (and the only reasonable solution I can think of). For methods that don't await things, the performance characteristics remain the same.

What do you think?

**Comment by kasbah** [comment]
I guess the issue with just going ahead and making them as async is that it encourages bad practices inside the method. It's a confusing API to the user since we give them an async function but then we actually block on it. 

The `schedule` function would at least encourage a good practice and set the right expectations about when the state will be modified. It would be a bit like `useEffect` in React. 

I'm not sure what the best way is actually. I'm a bit on the fence about this. 

--- 

On the side note about immutability. I see what you mean, but it would still be possible to work with that. I can convert the `Board` to JS primitives.

**Comment by hpx7** [comment]
Can you explain how the `schedule` function would work? Don't you have the same issue where the async callback inside schedule could be interleaved with execution from other methods?

**Comment by hpx7** [comment]
@kasbah I took a stab at it in https://github.com/hathora/hathora/pull/195

Even if the API might be confusing in that users could expect their methods not to block, at least having them block doesn't introduce any correctness issues. As for encouraging bad practices, I assume you mean it encourages users to blindly add awaits in methods without considering the throughput implications? This can be mitigated in part by documentation and later also by runtime analytics that can track method execution times and alert when things are taking long.

**Comment by kasbah** [comment]
> Can you explain how the schedule function would work? Don't you have the same issue where the async callback inside schedule could be interleaved with execution from other methods?
  
I was envisioning it being done on an idle loop, when there is nothing else to do. But yes, the way I modified the state in the example would require blocking too. One way around that could be to not allow modifying the state in the async function, but allow dispatching method calls 
  
 
  ```ts
  joinGame(
    state: InternalState,
    userId: UserId,
    ctx: Context,
    request: IJoinGameRequest
  ): Response {
    state.exampleUrl = `https://example.com/${request.exampleName}`;
    this.schedule(async (dispatch, state: readonly InternalState) => {
      const data = await fetch(state.exampleUrl).then((r) => r.json());
      dispatch(Method.SET_DATA, data as ISetDataRequest);
    });
    return Response.ok();
  }
  setData(
    state: InternalState,
    userId: UserId,
    ctx: Context,
    request: ISetDataRequest
  ): Response {
    state.data = request
    return Response.ok();
  }
```

Or with a bit of sugar, handing the user a sort of proxy of the `Impl` class instance: 
```ts
  joinGame(
    state: InternalState,
    userId: UserId,
    ctx: Context,
    request: IJoinGameRequest
  ): Response {
    state.exampleUrl = `https://example.com/${request.exampleName}`;
    this.schedule(async (proxy, state: readonly InternalState) => {
      const data = await fetch(state.exampleUrl).then((r) => r.json());
      proxy.setData(data as ISetDataRequest);
    });
    return Response.ok();
  }
...
```


Just marking the functions as async (like you are doing with #195) does seem a lot simpler. But yeah, I think it really does give the user the wrong impression about what they should be doing in the method. And they _will_ do http requests etc. You can't really combat that very well with just documentation.

**Comment by hpx7** [comment]
I think the dispatch approach works and is a nice suggestion, thanks.

The issue I see is that there is nothing that prevents the user modifying the original `state` arg from the method from inside the schedule callback.

I think both the `schedule` approach and the blocking approach have their pitfalls that need to be documented. At this time I'm leaning towards the blocking approach because (1) it's conceptually simpler for people to understand and (2) the downside of people's methods taking too long and resulting in degraded throughput seems less bad than the downside of introducing bugs by modifying state inside the `schedule` callback.

**Comment by kasbah** [comment]
Does the `readonly` not do the trick?

```ts
type AsyncCallback = (proxy, state: readonly InternalState) => Promise<void>;

function schedule(callback: AsyncCallback): void;
```

**Comment by hpx7** [comment]
What stops me from doing this:
```ts
  joinGame(state: InternalState,  userId: UserId, ctx: Context, request: IJoinGameRequest): Response {
    this.schedule(async () => {
      const data = await fetch(URL).then((r) => r.json());
      state.response = data;
    });
    return Response.ok();
  }
```

**Comment by kasbah** [comment]
Oh yeah, that is a problem (though I don't think your example will quite pass type checking). I'm still leaning towards some sort of solution that doesn't turn all that async into blocking calls. Seems like bad path for a web server. 

How about separating the effects out completely?

```ts
class Effects {
  async fetchData(state: readonly InternalState): Promise<void> {
     const data = await fetch(state.url)
     // sugar for something like this.impl.dispatch(Methods.SET_DATA, data as ISetDataRequest)
     this.impl.setData(data)
  }
}

class Impl {
   joinGame(...args) {
      // sugar for something like this.effects.dispatch(EffectMethods.FETCH_DATA)
      this.effects.fetchData()
      // ...
   }
   setData(state, userId, ctx, request) {
     state.data = request
     //...
   }
}

```

**Comment by hpx7** [comment]
Don't you usually _want_ to block?

```ts
  async setWeather(state: InternalState,  userId: UserId, ctx: Context, request: ISetWeatherRequest) {
    const response = await ctx.fetch<Weather>("http://weather.com/api/currentWeather");
    if (response.error) {
      return Response.error(response.error);
    }
    state.weather = response.json();
    return Response.ok();
  }
```
I would expect this method to simply run synchronously without interleaving any other methods I think?

**Comment by kasbah** [comment]
Conceptually you want to think about it like it's running sync. But practically, on a web server with multiple connections on a single thread, you don't want to block on http calls, file access, etc.

**Comment by hpx7** [comment]
The way I'm thinking about it is that you essentially want a mutex around the `stateId`. So the calls across game instances run in parallel, for a single instance everything is serial.

**Comment by kasbah** [comment]
Per `stateId` is a mitigation but if you are still blocking at least two connections with your awaits. The more players join one game, the worse it gets. 

The thing I'm still really stuck on with this is we would be essentially saying to people with the API: "sure, use `await` in the methods for some long running tasks". Whereas if they remain non-async it's really clear: "you can't expect to have your long running task return something right away, so do it elsewhere".

**Comment by hpx7** [comment]
You're right, maybe we need two modes of handling this. The first mode is for cases like my `setWeather` example where you don't expect the await to take too long and its much more natural to write in this style. The second mode would be for truly long running tasks where you have a way to dispatch method calls similar to how you suggested.

```ts
  async setWeather(state: InternalState, userId: UserId, ctx: Context, request: ISetWeatherRequest) {
    ctx.async(async (impl) => {
      const response = await ctx.fetch<Weather>("http://weather.com/api/currentWeather");
      impl.processWeatherResponse(response);
    });
    return Response.ok();
  }
  async processWeatherResponse(state: InternalState, userId: UserId, ctx: Context, request: ISetWeatherRequest) {
    if (request.error) {
      return Response.error(request.error);
    }
    state.weather = request.json;
    return Response.ok();
  }
  ```

It's less natural to write but it's non-blocking.

**Comment by kasbah** [comment]
I'm back on the fence. I don't think I can make this call at all :smile:. Your example has the same problem you pointed out before though. It doesn't prevent you (and kind of encourages you) to do:

```ts
 async setWeather(state: InternalState, userId: UserId, ctx: Context, request: ISetWeatherRequest) {
    ctx.async(async (impl) => {
      const response = await ctx.fetch<Weather>("http://weather.com/api/currentWeather");
      state.weather = response;
    });
    return Response.ok();
  }
  ```

**Comment by hpx7** [comment]
Yeah I think the specific api for this would have to be something like Redux
```ts
  async setWeather(state: InternalState, userId: UserId, ctx: Context, request: ISetWeatherRequest) {
    ctx.async(async (impl) => {
      const response = await ctx.fetch<Weather>("http://weather.com/api/currentWeather");
      ctx.dispatch(response, Method.PROCESS_WEATHER_RESPONSE);
    });
    return Response.ok();
  }
```

**Comment by kasbah** [comment]
That still has exactly the same problem though?

**Comment by hpx7** [comment]
Ha, yes you're right. Need to think more about what API would work

**Comment by kasbah** [comment]
Talking about redux, the problem of what we are trying to address is pretty similar to what [redux-saga](https://redux-saga.js.org/) tries to address. I never liked that generator based API personally... but it's a thing.

**Comment by kasbah** [comment]
Adapting my idea above to your example. Just food for thought. 

```ts
class AsyncMethods {
  async fetchWeather(state: readonly InternalState, ctx) {
    const response = await ctx.fetch<Weather>(`http://weather.com/api/currentWeather/${state.country}`);
    ctx.impl.processWeatherResponse(response);
  }
}

class Impl {
  setWeather(state: InternalState, userId: UserId, ctx: Context, request: ISetWeatherRequest) {
    state.country = request.country;
    ctx.async.fetchWeather();
    return Response.ok();
  }
  processWeatherResponse(state: InternalState, userId: UserId, ctx: Context, request) {
    if (request.error) {
      return Response.error(request.error);
    }
    state.weather = request.json;
    return Response.ok();
  }
}

```

**Comment by hpx7** [comment]
My concerns with the above example:
1. `setWeather` always returns `Response.ok()`, but the response should really depend on the result of the async call
2. `processWeatherResponse` looks like a method that is defined in `hathora.yml` and can be called by the client, but really it should be an internal method that is only defined in `impl.ts` and isn't exposed to the client
3. Passing arguments via modifying state (`state.country`) is kind of weird
4. It's hard to follow the logic in general

I think this addresses those concerns:
```ts
  setWeather(state: InternalState, userId: UserId, ctx: Context, request: ISetWeatherRequest) {
    if (getRoleForUser(userId) !== Role.ADMIN) {
      return Response.error("Only admins can update weather");
    }
    return ctx.async(async (state) => {
      const response = await ctx.fetch<Weather>(`http://weather.com/api/currentWeather/${request.country}`);
      if (response.error) {
        return Response.error(response.error);
      }
      state.weather = reponse.json();
      return Response.ok();
    });
  }
```

But of course this introduces two new challenges:
1. How do you prevent using the top level `state` inside `ctx.async` (maybe have a linter built into the cli that errors when it detects that?)
2. How do you chain multiple async calls inside a method

**Comment by hpx7** [comment]
Thinking about this more, I think the cleanest way to do this will be to use immutable internal state + [Optimistic Concurrency Control](https://en.wikipedia.org/wiki/Optimistic_concurrency_control). This would enable you to use `await` freely since you can be sure that `state` hasn't mutated after control returns.

Hathora already knows how to retry transactions safely so I think this would work nicely.

So now the challenge is making internal state immutable -- ill start a discussion for this in another issue.

**Comment by kasbah** [comment]
Ah, very interesting, I'm wondering if we should fix and merge #195 and then do OCC and immutability as an optimization later on? 

As for my specific use case, so I can stick with standard Hathora, I'm just going to do something that I know is technically wrong but probably won't cause any problems. I just call `then` on the promise I need and modify the state in the callback: https://github.com/kasbah/calm-go/blob/527a8850a6846f6da74edfb26c2f1d1878914d9c/server/impl.ts#L238-L246

**Comment by francislavoie** [comment]
Dang, this might end up being a blocker for me. My plan was to make HTTP requests in response to user actions to perform permanent state changes on my backend server.

What if `hathora.yml` could declare `asyncMethods:` that would have a signature like:

```
async methodName(state: Readonly<InternalState>, userId: UserId, ctx: Context, request: IMethodNameRequest): Response
```

And then any state changes would need to be performed via `ctx.stateChange((state) => { state.prop++ })` where the callback would always be synchronous, but you only get access to writable state within that callback and the current state via method params is readonly.

Edit: Well, `Readonly<T>` only deals with first-level properties, it doesn't prevent `state.prop.push()` and such. There's a bunch of `DeepReadonly<T>` solutions out there that might be appropriate though. Still not super clean.
